### PR TITLE
✨ React Server Component script

### PR DESCRIPTION
**Pull request by HofmannZ**
This PR introduces an alternative way to expose environment variables to the client.

Checklist:

- [x] Implement `EnvScript` rsc that sets env received from props.
- [x] Implement `PublicEnvScript` rsc that reads public env from `process.env` and passes it to the `EnvScript`.
- [x] Re-introduce utility from v1 that reads env from `window` or `process`.
- [x] Add/update tests.
- [x] Add/update example.
- [x] Add/update docs. 

Reading nonce from headers is blocked by: https://github.com/vercel/next.js/pull/58129

### CODE REVIEW

File: src/script/env-script.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * available, it will be set on the script tag.
+ *
+ * Usage:
+ * ```ts
+ * <head>
+ *   <EnvScript env={{ NODE_ENV: 'test', API_URL: 'http://localhost:3000' }} />
+ * </head>
+ * ```
+ */
+export const EnvScript: FC<EnvScriptProps> = ({ env }) => {
```
Comment (rfviolato) [inline src/script/env-script.tsx:22]: Would be nice to receive the `nonce` via props and mark it as optional, probably most of people won't have CSP enabled.

File: src/script/env-script.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  // It's strongly recommended to set a nonce on your script tags. For
+  // const nonce = headers().get('x-nonce') ?? undefined;
+
+  return (
+    <script
+      data-testid="env-script"
+      // nonce={nonce}
+      dangerouslySetInnerHTML={{
+        __html: `window.__ENV = ${JSON.stringify(env)}`,
```
Comment (rfviolato) [inline src/script/env-script.tsx:31]: You could also consider making the key to the environment variables a shared constant.

```tsx
import { PUBLIC_ENV_KEY } from 'contants.ts';

// ...
  return (
    <script
      data-testid="env-script"
      // nonce={nonce}
      dangerouslySetInnerHTML={{
        __html: `window[${PUBLIC_ENV_KEY}] = ${JSON.stringify(env)}`,
```

**Review by rfviolato — COMMENTED** [review]
Added a couple of remarks but LGTM regardless

File: src/script/env-script.tsx (hunk 1 above)
Comment (HofmannZ) [reply src/script/env-script.tsx]: Nice suggestion, I've also added the option to configure the nonce header key.

**Pull request merged**